### PR TITLE
Add triu tril and tri functions

### PR DESCRIPTION
**Pull request by msluszniak**
This one will be helpful in K-fold implementation.

### CODE REVIEW

**Comment by josevalim** [comment]
From my side, we should move this to `Nx.LinAlg`, we need to make sure it works as if it was batched for rank > 2 (including tests) and it raises for rank < 2.

But I would like to hear @polvalente thoughts before. :)

**Comment by msluszniak** [comment]
> From my side, we should move this to `Nx.LinAlg`, we need to make sure it works as if it was batched for rank > 2 (including tests) and it raises for rank < 2.
> 
> But I would like to hear @polvalente thoughts before. :)

Yeah, we definitely need to add tests for batched tensors and rank checks. I'm not sure about `Nx.LinAlg`.

**Comment by polvalente** [comment]
To elaborate more, I think they belong in `Nx` instead of `Nx.LinAlg` because they are just used for data masking. There are lots of Lin. Alg. operations that use them, but they aren't specific definitions of Lin. Alg. itself. Also, `Nx.eye` is a linear algebra definition that is present outside of Nx.LinAlg.

**Comment by josevalim** [comment]
Alright. So let’s go with the changes above and merge it.

File: nx/lib/nx.ex [hunk 1]
```diff
@@ -1391,6 +1391,117 @@ defmodule Nx do
...
+          [4, 5, 6],
+          [7, 8, 9]
+        ]
+      >
+  """
+  @doc type: :creation
+  def tril(tensor, opts \\ []) do
+    opts = keyword!(opts, k: 0)
+    mask = tri(axis_size(tensor, -2), axis_size(tensor, -1), k: opts[:k])
+    select(equal(mask, Nx.tensor(1, type: type(mask))), tensor, Nx.tensor(0, type: type(tensor)))
```
Comment (polvalente) [inline nx/lib/nx.ex:1427]:
```suggestion
    select(mask, tensor, Nx.tensor(0, type: type(tensor)))
```

File: nx/lib/nx.ex [hunk 2]
```diff
@@ -1391,6 +1391,117 @@ defmodule Nx do
...
+          [0, 0, 6],
+          [0, 0, 0]
+        ]
+      >
+  """
+  @doc type: :creation
+  def triu(tensor, opts \\ []) do
+    opts = keyword!(opts, k: 0)
+    mask = tri(axis_size(tensor, -2), axis_size(tensor, -1), k: opts[:k] - 1)
+    select(equal(mask, Nx.tensor(1, type: type(mask))), Nx.tensor(0, type: type(tensor)), tensor)
```
Comment (polvalente) [inline nx/lib/nx.ex:1463]:
```suggestion
    select(mask, 0, tensor)
```

File: nx/lib/nx.ex [hunk 3]
```diff
@@ -1391,6 +1391,117 @@ defmodule Nx do
...
+          [1, 1, 0],
+          [1, 1, 1],
+          [1, 1, 1]
+        ]
+      >
+  """
+  @doc type: :creation
+  def tri(n, m, opts \\ []) do
+    opts = keyword!(opts, k: 0)
+    greater_equal(new_axis(iota({n}), 1), new_axis(subtract(iota({m}), opts[:k]), 0))
```
Comment (polvalente) [inline nx/lib/nx.ex:1502]:
```suggestion
    greater_equal(iota({n, 1}), subtract(iota({0, m}), opts[:k]))
```

File: nx/lib/nx.ex [hunk 4]
```diff
@@ -1391,6 +1391,117 @@ defmodule Nx do
     eye(Nx.shape(tensor), opts)
   end
 
+  @doc """
+  Lower triangle of an array.
```
Comment (josevalim) [inline nx/lib/nx.ex:1395]:
```suggestion
  Lower triangle of a matrix.
```

And we also need to implement and document the batching behaviour. :)

File: nx/lib/nx.ex [hunk 5]
```diff
@@ -1391,6 +1391,182 @@ defmodule Nx do
...
+            [4, 5, 0, 0],
+            [8, 9, 10, 0]
+          ],
+          [
+            [12, 0, 0, 0],
+            [16, 17, 0, 0],
+            [20, 21, 22, 0]
+          ]
+        ]
+      >
```
Comment (josevalim) [inline nx/lib/nx.ex:1438]:
```suggestion
      >

```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**